### PR TITLE
feat: added bootnodes list as fallback mechanism

### PR DESCRIPTION
**Pull request by leandroyalet**
Refs #19

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -16,6 +16,58 @@ const rl = createInterface({
...
+  })
+}
+
+async function processBootnodes (node: any, bootNodeAddresses: Array<string>) {
+  let connected = false
+  const iterator = bootNodeAddresses[Symbol.iterator]()
+  let count = 0
+  let bootNodeAddr = iterator.next().value
+
+  while (count < bootNodeAddresses.length && !connected) {
```
Comment (vojtechsimetka) [inline src/index.ts]: If you want to mimic for cycle then use the `for` keyword :) . `for (let i = 0; i < bootNodeAddresses.length && !connected; ++i)`

File: src/index.ts [hunk 2]
```diff
@@ -16,6 +16,58 @@ const rl = createInterface({
...
+        }
+        resolve()
+      }
+    )
+  })
+}
+
+async function processBootnodes (node: any, bootNodeAddresses: Array<string>) {
+  let connected = false
+  const iterator = bootNodeAddresses[Symbol.iterator]()
```
Comment (vojtechsimetka) [inline src/index.ts]: I think @martinlarosa mystified you - making this an iterator breaks the readability - you can actually do that in the for cycle below if you want.

File: src/index.ts [hunk 3]
```diff
@@ -16,6 +16,58 @@ const rl = createInterface({
   output: process.stdout
 })
 
+function connectToBootnode (node: any, bootNodeAddr: string) {
+  return new Promise((resolve, reject) => {
+    async.parallel(
```
Comment (vojtechsimetka) [inline src/index.ts]: Here I would much rather if you used `Promise.race`. Some pattern like [this](https://italonascimento.github.io/applying-a-timeout-to-your-promises/) but I am actually sure there is something lower level that does timeouts on promises.

Comment (leandroyalet) [reply src/index.ts]: Yes, this is what I meant about to remove the _async_ library, I will try to use native functions instead 👌🏻

File: src/index.ts (hunk 1 above)
Comment (leandroyalet) [reply src/index.ts]: so C like 😄

Comment (vojtechsimetka) [reply src/index.ts]: yep :)

File: src/index.ts [hunk 4]
```diff
@@ -57,26 +109,9 @@ function mainProcess () {
...
-                logger.info('Connection Successful')
-
-                if (myArgs.chatClient) {
-                  const chatClient: CommandLineChat = new CommandLineChat(node)
-                  chatClient.init()
-                }
-              }
-            )
+          if (myArgs.bootNodeAddresses) {
+            const bootNodeAddresses: Array<string> = myArgs.bootNodeAddresses.split(',')
```
[the remark this answers is not in the document]
Comment (leandroyalet) [reply src/index.ts]: I changed that argument on yargs to an array

File: src/index.ts [hunk 5]
```diff
@@ -16,6 +16,43 @@ const rl = createInterface({
...
+  return Promise.race([
+    promise,
+    timeout
+  ])
+}
+
+function connectToBootnode (node: any, bootNodeAddr: (string | number)) {
+  logger.info(`Trying to connect to ${bootNodeAddr}`)
+  const dialPromise = new Promise<any>((resolve, reject) => {
+    node.dial(new Multiaddr(bootNodeAddr), (err: any, conn: any) => {
```
Comment (vojtechsimetka) [inline src/index.ts]: Isn't the `err` here of type `Error`?

File: src/index.ts [hunk 6]
```diff
@@ -16,6 +16,43 @@ const rl = createInterface({
...
+      } else {
+        resolve(conn)
+      }
+    })
+  })
+  return promiseTimeout(300, dialPromise)
+}
+
+function processBootnodes (node: any, bootNodeAddresses: (string | number)[]) {
+  return bootNodeAddresses.reduce((accumulatorPromise: Promise<any>, nextMultiAddress: (string | number)) => {
```
Comment (vojtechsimetka) [inline src/index.ts]: Can the `nextMultiAddress` be `number`?

File: src/index.ts (hunk 5 above)
Comment (leandroyalet) [reply src/index.ts]: yes! checked with the library

File: src/index.ts (hunk 6 above)
Comment (leandroyalet) [reply src/index.ts]: linter prank me... it is (string | Buffer | Multiaddr)

Comment (vojtechsimetka) [reply src/index.ts]: That is true, but if you look where it comes from (this is the `bootNodeAddresses` arg) it is a string only, right? :)

File: src/index.ts [hunk 7]
```diff
@@ -16,6 +17,85 @@ const rl = createInterface({
...
+
+/**
+ * Apply a timeout to a given promise.
+ *
+ * @param ms - timeout in milliseconds
+ * @param promise - Promise to fulfill or reject
+ *
+ * @returns Promise<any>
+ */
+const promiseTimeout = function (ms: number, promise: Promise<any>) {
```
Comment (vojtechsimetka) [inline src/index.ts]: Please make this a normal function

**Pull request merged**